### PR TITLE
PR: Add additional cases for code cells and block comments

### LINKED ISSUE #8: Some issues with special comments in the outline explorer
After PR spyder-ide/spyder#13885, a "special comment" is shown in the outline explorer, regardless of the number of consecutive `-` following the `#`.

Before the outline explorer migration to the pyls, a special comment was registered only when there was exactly 4 consecutive `-` following the `#`. I think this rule should be added back in the current implementation of this feature.

For example, the outline explorer is showing special comments for the horizontal separator that are sometimes present in the header of spyder's file.

image

## Versions

* Spyder version: 4.2.0.dev0 b7e9cf9b1
* Python version: 3.7.6 64-bit
* Qt version: 5.12.5
* PyQt5 version: 5.12.3
* Operating System: Windows 10

### LINKED ISSUE #9: Some cell patterns are not detected in the Outline explorer
## Problem Description

Use of cells denoted by #%% no longer displays in outline explorer
Use of comments denoted by #### no longer displays in outline explorer (even when option is checked)

### What steps reproduce the problem?

1.  Create source code with #%% notation of cells and/or #### comments
2.  Save code file
3.  Outline does not show cells or special comments

### What is the expected output? What do you see instead?

Expected output is the outline explorer shows cells and special comments in addition to functions

Actual output:  
image
image

### Paste Traceback/Error Below (if applicable)

```

NA

```

## Versions
```
* Spyder version: 4.2.0 None
* Python version: 3.7.4 64-bit
* Qt version: 5.12.9
* PyQt5 version: 5.12.3
* Operating System: Windows 10
```

### Dependencies
```

# Mandatory:
atomicwrites >=1.2.0            :  1.4.0 (OK)
chardet >=2.0.0                 :  3.0.4 (OK)
cloudpickle >=0.5.0             :  1.3.0 (OK)
diff_match_patch >=20181111     :  20181111 (OK)
intervaltree >=3.0.2            :  3.1.0 (OK)
IPython >=4.0                   :  7.19.0 (OK)
jedi =0.17.2                    :  0.17.2 (OK)
jsonschema >=3.2.0              :  3.2.0 (OK)
nbconvert >=4.0                 :  5.6.1 (OK)
numpydoc >=0.6.0                :  0.9.2 (OK)
paramiko >=2.4.0                :  2.7.2 (OK)
parso =0.7.0                    :  0.7.0 (OK)
pexpect >=4.4.0                 :  4.8.0 (OK)
pickleshare >=0.4               :  0.7.5 (OK)
psutil >=5.3                    :  5.7.3 (OK)
pygments >=2.0                  :  2.5.2 (OK)
pylint >=1.0                    :  1.9.4 (OK)
pyls >=0.36.1;<1.0.0            :  0.36.1 (OK)
pyls_black >=0.4.6              :  0.4.6 (OK)
pyls_spyder >=0.1.1             :  0.1.1 (OK)
qdarkstyle >=2.8                :  2.8.1 (OK)
qtawesome >=0.5.7               :  0.7.3 (OK)
qtconsole >=4.7.7               :  4.7.7 (OK)
qtpy >=1.5.0                    :  1.9.0 (OK)
setuptools >=39.0.0             :  41.0.0 (OK)
sphinx >=0.6.6                  :  1.8.5 (OK)
spyder_kernels >=1.10.0;<1.11.0 :  1.10.0 (OK)
three_merge >=0.1.1             :  0.1.1 (OK)
watchdog >=0.10.3               :  0.10.3 (OK)
zmq >=17                        :  19.0.2 (OK)

# Optional:
cython >=0.21                   :  0.29.14 (OK)
matplotlib >=2.0.0              :  3.3.2 (OK)
numpy >=1.7                     :  1.19.3 (OK)
pandas >=1.1.1                  :  1.1.4 (OK)
scipy >=0.17.0                  :  1.4.1 (OK)
sympy >=0.7.3                   :  None (NOK)

```

### PR DESCRIPTION
**Pull request by andfoy**
Fixes #8
Fixes #9

### CODE REVIEW

**Comment by andfoy** [comment]
@jnsebgosselin @eafpres @ccordoba12, could you please check if this PR address the missing code cells/block comments syntax?

**Comment by jnsebgosselin** [comment]
So based on what was said in PR https://github.com/spyder-ide/spyder/pull/13885, if I want to test that, I need to clone the repo locally, fetch the branch and install the repo with `pip install -e .` ?

Is there **any risk** that I'm going to break my conda installation by doing that? If I recall, Carlos always said that mixing pip and conda is something that is very bad to do.

**Comment by andfoy** [comment]
@jnsebgosselin, since this package does not depend on Spyder, it shouldn't break anything, so you can install it with pip without any problems, just make sure that is installed with `pip install -e . --no-deps`

**Comment by jnsebgosselin** [comment]
> 
> 
> @jnsebgosselin, since this package does not depend on Spyder, it shouldn't break anything, so you can install it with pip without any problems, just make sure that is installed with `pip install -e . --no-deps`

Ok excellent. The `--no-deps` part is good. I'm going to try it out later today.

**Comment by eafpres** [comment]
> @jnsebgosselin @eafpres @ccordoba12, could you please check if this PR address the missing code cells/block comments syntax?

Honestly I'm not sure I'm competent to test this without breaking something locally, so please do not wait on me.

**Comment by andfoy** [comment]
@eafpres Don't worry, just a quick question should `# ###` (note the space before the 3 #'s) be a block comment as well?

**Comment by eafpres** [comment]
> question should `# ###` (note the space before the 3 #'s) be a block comment as well?

I can understand that with the # %% case but this makes less sense to me. However, I’d rather defer to the dev team on that.

File: pyls_spyder/plugin.py
Comment (ccordoba12) [inline pyls_spyder/plugin.py]: Please add a comment above all these constants to mention what kind of cell each pattern detects.

File: pyls_spyder/plugin.py
Comment (ccordoba12) [inline pyls_spyder/plugin.py]: Please do the same for these constants.

File: pyls_spyder/plugin.py
Comment (ccordoba12) [inline pyls_spyder/plugin.py:111]: Why did you add `or cell_rule != CELL_PERCENTAGE` here?

File: pyls_spyder/utils.py
Comment (ccordoba12) [inline pyls_spyder/utils.py:15]: Please add a docstring to this class to understand what it does.

**Review by ccordoba12 — COMMENTED** [review]
This looks pretty good, thanks @andfoy!

File: pyls_spyder/plugin.py
Comment (andfoy) [reply pyls_spyder/plugin.py:111]: Because nesting is not well defined for the other kind of cells

Comment (ccordoba12) [reply pyls_spyder/plugin.py:111]: Right, good point.

**Comment by andfoy** [comment]
@ccordoba12, all the review comments should be fixed now

File: pyls_spyder/tests/test_plugin.py
Comment (ccordoba12) [inline pyls_spyder/tests/test_plugin.py]:
```suggestion
#%% Empty cell
```

This is to check that we also detect cells when there's no space between `#` and `%%` (which was the main regression reported in #9).

**Review by ccordoba12 — COMMENTED** [review]
Last comment for you @andfoy, then this should be ready.

**Review by ccordoba12 — APPROVED** [review]
Thanks @andfoy!

**Comment by MCilento93** [comment]
> @ccordoba12, all the review comments should be fixed now

Hello
Which release has the issue of review comments fixed?

**Comment by andfoy** [comment]
Hi @MCilento93, this fix is available in pyls-spyder v0.2.1, released yesterday

**Comment by MCilento93** [comment]
Mmm I have spyder 4.2.0 o 4.1.5

Can I install or do I need to wait for next spyder release?

Sincerely

**Comment by andfoy** [comment]
You can install it by using conda or pip:

```bash
# Anaconda (Recommended)
conda install pyls-spyder=0.2.1 -c spyder-ide

# Pip
pip install -U pyls-spyder
```

**Comment by MCilento93** [comment]
> You can install it by using conda or pip:
> 
> ```shell
> # Anaconda (Recommended)
> conda install pyls-spyder=0.2.1 -c spyder-ide
> 
> # Pip
> pip install -U pyls-spyder
> ```

installed, but unfortunately I do not see the comments row (like ### , !!! and so forth)

**Pull request merged**
